### DISCUSSION TITLE
Does caffeine can dynamicly adjustment cache size by JVM memory watermark?

### DISCUSSION THREAD

**Discussion by guhanjie** [ben-manes/caffeine#564] [Q&A] [answered]
hi, @ben-manes , I have read your answer at this [topic](https://stackoverflow.com/a/39503450), i agree with you.

I met the likely problem, i use caffeine to cache plan entry(200K per object) in a SQL engine(actually Presto), 
the cache policy is maximumSize=100000, my JVM memory is enough, will not OOM, but the cache entry will stay on Old Gen all the time, and it will lead to GC pressure.

I'm thinking about that, is there any GC-friendly policy for caffeine?
Does caffeine can dynamicly adjustment cache size by JVM memory watermark?

I used to realized it on hand, but the maximumSize seems to be  static initialized with Cache, can not be update at runtime.

Anything tips?

**Comment by ben-manes** [comment, marked as answer]
The cache does not monitor available memory to adjust any setting. If you wish to do so, then it can be changed dynamically using the `Cache.policy()` api.
```java
cache.policy().eviction().ifPresent(eviction -> {
  eviction.setMaximum(2 * eviction.getMaximum());
});
```

You can combine size-based eviction with soft references if you want a failsafe. This does add a small amount of per-entry [memory overhead](https://github.com/ben-manes/caffeine/wiki/Memory-overhead), though.

The cache's size eviction policy should be more GC friendly than a typical LRU, which violates the generational hypothesis (making it a GC benchmark for worst-case behavior). A SQL database is MRU / LFU biased, so the cache will auto-configure itself such that the admission policy will more readily reject a recent but infrequently used entry. As an MRU, this should increase the chance that the evicted entry is in the younger generation and reduce the GC pressure.

**Reply by guhanjie** [reply]
Thanks a lot

Cache.policy() is nice to me.

Your suggestion, "size-based eviction with soft references" is my first idea, but my scenario is latency sensitive, as you mentioned this [topic](https://stackoverflow.com/a/39503450), I worry about __Soft references__ will hurt performance in long-running JVMs. does it clear only on full GC??? if so, it will be too late, the heap pressure on that time will be too high, and GC time increased sharply, latency will be unbearable.

About eviction policy,
I choose caffeine, just because the Tiny LFU policy, rather than Guava's LRU, 
which I think LFU is more appropriate, because most of our SQL is periodic,
but a little different thing is, I will keep sql plan cache all the time until the plan is invalid(maybe meta info expired), so actually, the cache entry will live on Old Gen mustly. So I must take care the JVM memory watermark to avoid the GC pressure. Now, I'm going to use Cache.policy() to Dynamic Adjustment the maximum size according the value of ```Runtime.getRuntime().freeMemory() / Runtime.getRuntime().maxMemory()```

You also mentioned __MRU__, what's that?
caffeine's default eviction policy is TinyLFU, or MRU? I need to make sure it.

**Reply by ben-manes** [reply]
I agree that soft references should be avoided whenever possible, and only used if absolutely necessary. A modern GC is more tolerant to them by more aggressively discarding them, e.g. G1 is regional so each can be collected at different intervals. It may be more tolerant, but that gets into too many assumptions about the implementation details of a GC algorithm to rely upon.

MRU is the opposite of LRU, where it discards the most recently accessed entries. This is helpful in a workload that performs scans because those entries may never be observed again. In a loop, an LRU would always make the wrong choice as shown by the glimpse trace below.

![glimpse](https://raw.githubusercontent.com/ben-manes/caffeine/master/wiki/efficiency/glimpse.png)

Caffeine's policy is an adaptive W-TinyLFU algorithm, which allows it to obtain the best hit rates in any workload. By sampling the hit rates, it reconfigures the admission window size using a hill climber, thereby allowing it to give preference towards either LRU or TinyLFU. A stress test between strongly biased LRU and MRU workloads shows how it quickly corrects itself.

![adaptive](https://raw.githubusercontent.com/ben-manes/caffeine/master/wiki/efficiency/adaptive.png)

**Reply by guhanjie** [reply]
Great, thanks again :)

**Reply by krishna81m** [reply]
@guhanjie , could you share your caffeine policy to evict based on `Runtime.getRuntime().freeMemory() / Runtime.getRuntime().maxMemory()` as it is impossible to do that. When you start evicting, actual GC to trigger freeing memory can happen much much later.

@ben-manes , our Caffeine use case was hindered by the fact that our data ranges from 1 KB - 10 MB and a 10000 limit would range 10 MB - 100 GB causing _serious OOM issues_ as Caffeine does not support memory size based eviction. However, we do know data byte sizes to track total cache size. Are there _better policy() examples_ that allow us to trigger eviction based on _total cache size_ assuming we use expiryOnWrite and eviction listeners to keep track of size? Otherwise, we have to again switch back to good ol EhCache with custom refresh/expiry hooks :(.

**Reply by ben-manes** [reply]
@krishna81m In that case you would use `maximumWeight` and a `Weigher`. This measures the entry's size and bounds the capacity by that. The `maximumSize` setting simply uses a weigher where every entry has unit capacity. Unlike Ehcache, we do not specify how the weight is calculated (as that is magic, often wrong, and usually slow - see the [JEP](https://openjdk.java.net/jeps/8249196)). Let me know if that works for you.

**Reply by krishna81m** [reply]
Ouch, totally forgot about weigher, was trying to break my head understanding how to leverage Policy instead. This [example](https://github.com/apache/druid/blob/master/server/src/main/java/org/apache/druid/client/cache/CaffeineCache.java#L64) is what we needed. 

Just to clarify, both maximumSize and maximumWeight are not possible together, but, eviction/expiry/refresh by time and maximumWeight based evictions should still work together? And entries with higher weights that are more frequently used should not be evicted following default TinyLFU algorithm.

**Reply by ben-manes** [reply]
Yep.
- maximumSize and maximumWeight are exclusive to each other as maximumSize is shorthand for using `Weigher.singletonWeigher()`. Internally the cache is always weight-aware and optimizes out unnecessary fields.
- size/weight, expiration, and refresh policies can all be combined and will work together.
- TinyLFU will only consider the entry's popularity, so an entry with a higher frequency is preferred regardless of its weight. There is a [paper](https://arxiv.org/abs/2105.08770) evaluating taking weights into consideration, but we have not yet adopted its recommendations.

**Reply by guhanjie** [reply]
> @guhanjie , could you share your caffeine policy to evict based on `Runtime.getRuntime().freeMemory() / Runtime.getRuntime().maxMemory()` as it is impossible to do that. When you start evicting, actual GC to trigger freeing memory can happen much much later.

yes, GC would do free memory later, but if we make sure the some of cache has no reference(invalid by Caffeine, and not refer in any your own code), the GC will be easy to free them once triggered.

So, my policy is just simple to limit the cache size by monitoring jvm mem, not the strict limit, but the test is effective.
```java
private final double USED_MEM_WATERMARK_RATIO = 0.75;
....        
scheduler.scheduleAtFixedRate(() -> {
            planCache.policy().eviction().ifPresent(eviction -> {
                long used = memoryUsage.getUsed();
                long total = memoryUsage.getMax();
                double ratio = used*1.0/total;
                if(ratio > USED_MEM_WATERMARK_RATIO) {
                    lastUpdateTime.set(System.currentTimeMillis());
                    long size = eviction.getMaximum()/2;
                    eviction.setMaximum(size); //mem watermark high, shrink cache size
                    log.info("Plan cache shrink, setMaximum %d, jvm memory water high: %d(used)/%d(total)", size, used, total);
                }
                else {
                    long durationInMills = System.currentTimeMillis()-lastUpdateTime.get();
                    if(eviction.getMaximum() < maximumSize && durationInMills > gracePeriodInMillis) {
                        lastUpdateTime.set(System.currentTimeMillis());
                        long size = Math.max(eviction.getMaximum()*2, maximumSize);
                        eviction.setMaximum(size); //mem watermark low, expand cache size
                        log.info("Plan cache expand, setMaximum %d, jvm memory water low: %d(used)/%d(total)", size, used, total);
                    }
                }
            });
        },5, 5, TimeUnit.SECONDS);
```

**Reply by adityamohan93** [reply]
@ben-manes Any chance the value level weight aware eviction is in the works? For our use case, we cache our database entries with String key and Long list value using Caffeine and there are usually queries that are not available in the database that results in added lookup time(ser-de) as well as database resource usage. We don't cache these unavailable queries as keys with something like an empty list in our Caffeine cache but considering it now. However it would lead to larger cache size as the String keys would be saved regardless and we might end up evicting non-empty lists prematurely. TinyLFU works on frequency, however if there is a weight factor also coupled with it, we can assign weight as zero to these empty cached lists so that they get evicted first if frequencies are similar between a zero and non-zero weighted key.

**Reply by ben-manes** [reply]
@adityamohan93,

A zero weighted key has an explicit meaning of don't size evict as it takes no capacity (used for pinning entries). So you could use negative caching without causing size evictions, but might also want to use an expiration policy to discard them to avoid runaway growth.

You shouldn't need to make business logic assumptions based on how the algorithm works. That's shared because it is fun CS stuff and I enjoy learning from the community's feedback. The size-aware TinyLFU only adds a modest benefit in the paper so you should still get excellent hit rates with our current approach. I also think your ideal might not work even then because we select by recency order (LRU lists) whereas you might be imagining an ordered queue based on a scoring function.
